src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                }
+                if (result == ConsumeTokenResult.IncompleteRollback)
+                {
+                    _consumed = prevConsumed;
+                    _tokenType = prevTokenType;
+                    _lineBytePosition = prevPosition;
+                    _lineNumber = prevLineNumber;
+                }
+                goto Done;
+            }
+
+            retVal = true;
+
+        Done:
+            return retVal;
+
+        ReadFirstToken:
+            retVal = ReadFirstToken(first);
+            goto Done;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData()
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (_isNotPrimitive && IsLastSpan)
+                {
+                    if (_tokenType != JsonTokenType.EndArray && _tokenType != JsonTokenType.EndObject)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidEndOfJson);
+                }
+                return false;
+            }
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool HasMoreData(ExceptionResource resource)
+        {
+            if (_consumed >= (uint)_buffer.Length)
+            {
+                if (IsLastSpan)
+                    ThrowHelper.ThrowJsonReaderException(ref this, resource);
+                return false;
+            }
+            return true;
+        }
+
+        private bool ReadFirstToken(byte first)
+        {
+            if (first == JsonConstants.OpenBrace)
+            {
+                _currentDepth++;
+                _stackFreeContainer = 1;
+                _tokenType = JsonTokenType.StartObject;
+                _consumed++;
+                _lineBytePosition++;
+                _inObject = true;
+                _isNotPrimitive = true;
+            }
+            else if (first == JsonConstants.OpenBracket)
+            {
+                _currentDepth++;
+                _tokenType = JsonTokenType.StartArray;
+                _consumed++;
+                _lineBytePosition++;
+                _isNotPrimitive = true;
+            }
+            else
+            {
+                //Create local copy to avoid bounds checks.
+                ReadOnlySpan<byte> localCopy = _buffer;
+
+                if ((uint)(first - '0') <= '9' - '0' || first == '-')
+                {
+                    if (!TryGetNumber(localCopy.Slice(_consumed), out int numberOfBytes))
+                        return false;
+                    _tokenType = JsonTokenType.Number;
+                    _consumed += numberOfBytes;
+                    _lineBytePosition += numberOfBytes;
+                    goto Done;
+                }
+                else if (ConsumeValue(first))
+                    goto Done;
+
+                return false;
+
+            Done:
+                if (_consumed >= (uint)localCopy.Length)
```

Comment:
It'd be good to have a comment explaining why this isn't just calling HasMoreData.